### PR TITLE
refactor: Upload use origin behavior

### LINKED ISSUE #29707: antd v4.13.1 upload 自定义上传，会执行2次
- [ ] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[https://riddle.alibaba-inc.com/riddles/e169534a](https://riddle.alibaba-inc.com/riddles/e169534a)

### Steps to reproduce
antd v4.13.1 Upload组件 自定义上传方法，会执行2次

### What is expected?
执行一次

### What is actually happening?
执行两次

| Environment | Info |
|---|---|
| antd | 4.13.1 |
| React | 16.14.0 |
| System | MacOS |
| Browser | Chrome |

### PR DESCRIPTION
**Pull request by zombieJ**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

resolve #29707
resolve #29706

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

**Comment by kerm1it** [comment]
是把 Proxy 去掉了么？

**Comment by zombieJ** [comment]
嗯，https://github.com/ant-design/ant-design/pull/29718#issuecomment-796787160 。到 antd 5 里再去掉这 `beforeUpload` 特殊情况。

File: components/upload/Upload.tsx [hunk 1]
```diff
@@ -120,29 +153,38 @@ const InternalUpload: React.ForwardRefRenderFunction<unknown, UploadProps> = (pr
...
-    targetItem.status = 'uploading';
-
-    const nextFileList = replaceFileList(targetItem, mergedFileList);
+    objectFileList.forEach((fileObj, index) => {
+      // Repeat trigger `onChange` event for compatible
+      let triggerFileObj: UploadFile = fileObj;
+
+      if (!filteredFileInfoList[index].parsedFile) {
+        // `beforeUpload` return false
+        const { originFileObj } = fileObj;
```
Comment (kerm1it) [inline components/upload/Upload.tsx:172]: 这里既然 beforeUpload 已经返回 false，为啥还要触发 onChange？

Comment (zombieJ) [reply components/upload/Upload.tsx:172]: `false` 仍然会添加 item 到列表中，这边是受控需要。

File: components/upload/Upload.tsx [hunk 2]
```diff
@@ -120,29 +153,38 @@ const InternalUpload: React.ForwardRefRenderFunction<unknown, UploadProps> = (pr
...
+        const clone = (new File([originFileObj], originFileObj.name, {
```
Comment (kerm1it) [inline components/upload/Upload.tsx:173]: 这里的克隆看着有点奇怪，其他的没问题了。

Comment (zombieJ) [reply components/upload/Upload.tsx:173]: 嗯，为了之后 v5 都用 `info.file.originFileObj` 提前做准备。clone 免得 `originFileObj` 嵌套循环了~

Comment (kerm1it) [reply components/upload/Upload.tsx:173]: 明白，就是上次的 proxy 被循环嵌套

**Review by kerm1it — APPROVED** [review]

File: components/upload/interface.tsx [hunk 3]
```diff
@@ -25,7 +26,7 @@ export interface UploadFile<T = any> {
   status?: UploadFileStatus;
   percent?: number;
   thumbUrl?: string;
-  originFileObj?: File | Blob;
+  originFileObj: RcFile;
```
Comment (afc163) [inline components/upload/interface.tsx:29]: originFileObj 为啥是 RcFile？

Comment (kerm1it) [reply components/upload/interface.tsx:29]: originFileObj 拿的其实就是从 RC 传过来的，继承于 File，额外加了 uid 字段

Comment (zombieJ) [reply components/upload/interface.tsx:29]: 之前 `beforeUpload` 除了 `File` 和 `Blob` 外还可以把它改成 `string` 类型，我们类型其实一直都不太对。这边帮用户直接内部封装成最高类型 RcFile（RcFile 继承 File 继承 Blob），也不用判断是 File 还是 Blob 了。

**Review by afc163 — APPROVED** [review]

**Pull request merged**